site/docs/hosting/vps.md

```diff
@@ -52,130 +152,616 @@ systemd is a powerful service manager which is pre-installed on many Linux distr
+#
+# Unless the file starts with a global options block, the first
+# uncommented line is always the address of your site.
+#
+# To use your own domain name (with automatic HTTPS), first make
+# sure your domain's A/AAAA DNS records are properly pointed to
+# this machine's public IP, then replace ":80" below with your
+# domain name.
+
+:80 {
+  # Set this path to your site's directory.
+  root * /usr/share/caddy
+
+  # Enable the static file server.
+  file_server
+
+  # Another common task is to set up a reverse proxy:
+  # reverse_proxy localhost:8080
+
+  # Or serve a PHP site through php-fpm:
+  # php_fastcgi localhost:9000
+}
+
+# Refer to the Caddy docs for more information:
+# https://caddyserver.com/docs/caddyfile
+```
+
+For the bot to work, make the configuration look like this:
+
+```text
+<domain> {
+  reverse_proxy /<token> localhost:<port>
+}
 ```
 
-### Managing Apps
+> Replace `<domain>` with your domain, `<token>` with your bot token, and `<port>` with the port on which you want to run your bot.
 
-#### Start
+Reload Caddy every time you change the site's configuration file using the following command:
 
 ```sh
-pm2 start --name <app_name> <entry_point>
+systemctl reload caddy
 ```
 
-> The `<app_name>` can be any identifier to your app, for example: `bot1`.
-> The `<entry_point>` should be the path to your index file (which runs your bot).
+Now all requests to the address `https://<domain>/<token>` will be redirected to the address `http://localhost:<port>/<token>`, where the bot's webhook is running.
 
-#### Restart
+#### Connecting a Webhook to Telegram
 
-By restarting, you stop the app, and start it again.
+All you have to do is tell Telegram where to send the updates.
+To do this, open your browser and visit the page at the following link:
 
-```sh
-pm2 restart <app_name>
+```text
+https://api.telegram.org/bot<token>/setWebhook?url=https://<domain>/<token>
 ```
 
-#### Reload
+> Replace `<token>` with your bot token and `<domain>` with your domain.
+
+## CI/CD
+
+[CI/CD](https://about.gitlab.com/topics/ci-cd) is an important part of the modern software development process.
+This guide covers almost all of the [CI/CD pipeline](https://about.gitlab.com/topics/ci-cd/cicd-pipeline).
+
+We will focus on writing scripts for GitHub and GitLab.
+You can easily adapt the examples below to your CI/CD service of choice, such as Jenkins, Buddy, etc., if needed.
 
-By reloading, you replace the current process of your app with a new one, resulting in a 0-second downtime.
-This is recommended for stateless applications.
+:::tip Self-hosted Runner
+GitHub and GitLab offer a certain amount of resources for free to complete your tasks.
+However, when setting up your pipeline, you may quickly use them all up, which will require you to pay money for additional resources or the tasks will not run.
+To prevent this from happening, we recommend installing a self-hosted runner on your server so that tasks run on your server.
```

Comment:
Do we recommend that? I usually just open-source the bots, which gives you way more CI time than you need. Isn't that a better option? We can still recommend hosting your own runner if publishing the source is not an option, but doing this isn't exactly trivial, so I wouldn't blindly tell everyone to do it.

Reply:
Self-hosted runner has nothing to do with the openness of the bot's source code.

Reply:
In any case, we can leave a "hint" instead of a "recommendation" about the possibility of installing your own runner. For example, "to prevent this from happening, ~~we recommend to~~ **you can** install a self-hosted runner on the server".

Reply:
Self-hosted runners do not have the same restrictions as official runners. Self-hosted runners can also be used for closed projects. The main thing is to have enough resources on the server to support your own runner.

Reply:
I know. The point was this:

Problem: not enough CI resources on your private project
Solution 1: make it public
Solution 2: host a runner

Solution 1 is much easier, yet we currently recommend solution 2.

Reply:
No. In my opinion, the reason and solution are as follows:
Reason: the free plan of the official CI/CD is not enough for large projects or even for small ones if the user is experimenting with CI/CD for the first time and performing a lot of pipelines in a short time.
Solution 1: pay for additional CI/CD resources, which can be very expensive, as in the case of cloud hosting.
Solution 2: Install your own runner on the server to be limited only by the capacity of your own server, which can be upgraded at any time. (In fact, even self-hosted runners have limitations, but they are much smaller then free ones). In this case, it doesn't matter if the bot's source code is open source; no additional restrictions will be imposed. And there is no point in opening access to your own runner either.